### ISSUE TITLE
[R-package] construct yields different results

### ISSUE THREAD

**Issue by Laurae2** [Microsoft/LightGBM#800]
## Environment info
Operating System: Windows 8.1 Pro
CPU: i7-4600U
C++/Python/R version: 3.4.0

When using `construct`, training metric differs. It is 100% reproducible.

## Steps to reproduce

1. Run this (without construct):

```r
lgb.unloader(wipe = TRUE)
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label, weight = inverse.rle(list(lengths = c(rep(500, 12), 513), values = 1:13)), init_score = inverse.rle(list(lengths = c(rep(500, 12), 513), values = (1:13) * 2)))
params <- list(objective = "regression", metric = "l2")
# dtrain$construct()

model <- lgb.train(params,
                   dtrain,
                   10,
                   list(train = dtrain),
                   min_data = 1,
                   learning_rate = 1,
                   early_stopping_rounds = 10,
                   reset_data = TRUE)

# Check whether data is copied deeply or not
dtrain$setinfo("weight", inverse.rle(list(lengths = c(rep(500, 12), 513), values = 13:1)))

# Not identical training! // CORRECT
model <- lgb.train(params,
                   dtrain,
                   10,
                   list(train = dtrain),
                   min_data = 1,
                   learning_rate = 1,
                   early_stopping_rounds = 10)
```

2. Run this (with construct) and compare:

```r
lgb.unloader(wipe = TRUE)
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label, weight = inverse.rle(list(lengths = c(rep(500, 12), 513), values = 1:13)), init_score = inverse.rle(list(lengths = c(rep(500, 12), 513), values = (1:13) * 2)))
params <- list(objective = "regression", metric = "l2")
dtrain$construct()

model <- lgb.train(params,
                   dtrain,
                   10,
                   list(train = dtrain),
                   min_data = 1,
                   learning_rate = 1,
                   early_stopping_rounds = 10,
                   reset_data = TRUE)

# Check whether data is copied deeply or not
dtrain$setinfo("weight", inverse.rle(list(lengths = c(rep(500, 12), 513), values = 13:1)))

# Not identical training! // CORRECT
model <- lgb.train(params,
                   dtrain,
                   10,
                   list(train = dtrain),
                   min_data = 1,
                   learning_rate = 1,
                   early_stopping_rounds = 10)
```

---

Run 1:

```r
[1]:	train's l2:2.56938 
[2]:	train's l2:2.50057 
[3]:	train's l2:2.41568 
[4]:	train's l2:2.37591 
[5]:	train's l2:2.34725 
[6]:	train's l2:2.30721 
[7]:	train's l2:2.26781 
[8]:	train's l2:2.21442 
[9]:	train's l2:2.17173 
[10]:	train's l2:2.12131 

[1]:	train's l2:2.84081 
[2]:	train's l2:2.68119 
[3]:	train's l2:2.62447 
[4]:	train's l2:2.55463 
[5]:	train's l2:2.49118 
[6]:	train's l2:2.44908 
[7]:	train's l2:2.37546 
[8]:	train's l2:2.33494 
[9]:	train's l2:2.29519 
[10]:	train's l2:2.25349
```

Run 2:

```r
[1]:	train's l2:2.56938 
[2]:	train's l2:2.4835 
[3]:	train's l2:2.41925 
[4]:	train's l2:2.36668 
[5]:	train's l2:2.31415 
[6]:	train's l2:2.2731 
[7]:	train's l2:2.24578 
[8]:	train's l2:2.21683 
[9]:	train's l2:2.19564 
[10]:	train's l2:2.16324 

[1]:	train's l2:2.84081 
[2]:	train's l2:2.67998 
[3]:	train's l2:2.62326 
[4]:	train's l2:2.55343 
[5]:	train's l2:2.4909 
[6]:	train's l2:2.44912 
[7]:	train's l2:2.385 
[8]:	train's l2:2.33234 
[9]:	train's l2:2.2651 
[10]:	train's l2:2.22048 
```

**Comment by guolinke** [comment]
@Laurae2 min_data will affect the dataset construction. 
And you didn't pass it as the parameter of Dataset.

**Comment by Laurae2** [comment]
@guolinke It now matches when passing `min_data = 1` to `construct`.

Do you know the list of Dataset-only parameters?

**Comment by guolinke** [comment]
@Laurae2 
Check this https://github.com/Microsoft/LightGBM/blob/master/include/LightGBM/config.h#L94-L153
